### PR TITLE
[Codegen] Add `DeviceMappingAttr` that maps to workgroup IDs.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
This adds a `DeviceMappingAttr` that maps to
`hal.interface.workgroup.id`s, but for use as mapping attribute with `scf.forall`. The existing attributes upstream dont seem to match the behavior one would expect for mapping to GPU dimensions. Overtime this should be upstreamed to MLIR (and deprecate what exists today).

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/test/workgroup_mapping_attrs.mlir [hunk 1]
```diff
@@ -0,0 +1,37 @@
+// RUN: iree-opt --split-input-file --verify-diagnostics --allow-unregistered-dialect %s | FileCheck %s
+
+func.func @roundtrip() {
+  "dummy.op"() {
+    workgroup_mapping = [
+      #iree_codegen.workgroup_mapping<x>,
+      #iree_codegen.workgroup_mapping<y>,
+      #iree_codegen.workgroup_mapping<z:0>,
+      #iree_codegen.workgroup_mapping<z:1>]
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/test/workgroup_mapping_attrs.mlir:9]: Oh cool, this is a nice way to do it. Much better than the relative ID thing upstream.

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp [hunk 2]
```diff
@@ -363,6 +363,90 @@ CompilationInfoAttr::verify(function_ref<InFlightDiagnostic()> emitError,
...
+                                               WorkgroupId id) {
+  return WorkgroupMappingAttr::get(context, id, /*delinearizedDim=*/0);
+}
+
+LogicalResult
+WorkgroupMappingAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+                             WorkgroupId id, int64_t delinearizedDim) {
+  if (delinearizedDim > 0 && id != WorkgroupId::IdZ) {
+    return emitError() << "illegal to use `delinearizationDim` for "
+                       << stringifyWorkgroupId(id);
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp:380]: why is delinearization only legal for `z`?

**Review by qedawkins — COMMENTED** [review]
I like the design of the new attribute! LGTM so far but need to take a closer look a bit later.

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp:380]: IMO, the x, y, z itself is a dilinearization of ids. So I dont see much value in have delinearization of x and y separately. You need delinearization on z to be able to map "logical 4D" to these three dimensions. Also makes the resolution logic easier, cause you handle x and y in a particular way and any delinearization you need to do on top of that just map that to Z. I have only seen this usage in practice.

Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp:380]: ah I see, that makes sense because the result of workgroup tiling sets the sizes. I guess the same thing doesn't work the same way on GPU because there might be multiple different logical bases (i.e. multiple different scf.forall ops), but maybe it would be a good idea to try to set the actual 3D workgroup size later in the pipeline and only set the flat 1D workgroup size in KernelConfig.

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp [hunk 3]
```diff
@@ -363,6 +363,90 @@ CompilationInfoAttr::verify(function_ref<InFlightDiagnostic()> emitError,
...
+  }
+  return success();
+}
+
+LogicalResult WorkgroupMappingAttr::verifyAttrList(
+    MLIRContext *context, function_ref<InFlightDiagnostic()> emitError,
+    ArrayRef<Attribute> attrs) {
+  if (attrs.empty()) {
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp:388]: What does an empty mapping list mean?

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp [hunk 4]
```diff
@@ -363,6 +363,90 @@ CompilationInfoAttr::verify(function_ref<InFlightDiagnostic()> emitError,
...
+  for (auto [index, attr] : llvm::enumerate(mappingAttrsRef.drop_front(2))) {
+    if (attr.getDelinearizedDim() != index) {
+      return emitError() << "missing `workgroup_mapping<z:" << index;
+    }
+  }
+  return success();
+}
+
+int64_t WorkgroupMappingAttr::getMappingId() const {
+  return static_cast<int64_t>(getId()) + getDelinearizedDim();
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp]: nit: you can use `llvm::to_underlying(getId())`

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.td [hunk 5]
```diff
@@ -126,6 +131,73 @@ def DispatchLoweringPassPipelineAttr :
...
+  let mnemonic = "workgroup_mapping";
+
+  let parameters = (ins
+    EnumParameter<WorkgroupIdEnum>:$id,
+    DefaultValuedParameter<"int64_t", "0">:$delinearizedDim
+  );
+  let assemblyFormat = "`<` $id (`` `:` `` $delinearizedDim^)? `>`";
+
+  let description = [{
+    Attribute that evventually will be used to map distributed loop iterations
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.td]:
```suggestion
    Attribute that eventually will be used to map distributed loop iterations
```

**Review by qedawkins — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp:380]: maybe.... I dont think I fully grok all of that, but if you have an example handy it will help me anchor what you are saying.

File: compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp (hunk 3 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.cpp:388]: its a corner case, and I was trying to think about whether I should fail or pass. For now I am just ignoring but it shouldnt happen. 
Really I cant test the list anywhere. Subsequent PRs will use this method. Ill revisit if empty list needs to be handled differently.
